plugins/mapper-annotated-text/src/main/java/org/elasticsearch/index/mapper/annotatedtext/AnnotatedTextFieldMapper.java

```diff
@@ -0,0 +1,789 @@
+            AnnotatedTextFieldMapper.Builder builder = new AnnotatedTextFieldMapper.Builder(fieldName);
+            
+            builder.fieldType().setIndexAnalyzer(parserContext.getIndexAnalyzers().getDefaultIndexAnalyzer());
+            builder.fieldType().setSearchAnalyzer(parserContext.getIndexAnalyzers().getDefaultSearchAnalyzer());
+            builder.fieldType().setSearchQuoteAnalyzer(parserContext.getIndexAnalyzers().getDefaultSearchQuoteAnalyzer());
+            parseTextField(builder, fieldName, node, parserContext);
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals("position_increment_gap")) {
+                    int newPositionIncrementGap = XContentMapValues.nodeIntegerValue(propNode, -1);
+                    builder.positionIncrementGap(newPositionIncrementGap);
+                    iterator.remove();
+                } else if (propName.equals("eager_global_ordinals")) {
+                    builder.eagerGlobalOrdinals(XContentMapValues.nodeBooleanValue(propNode, "eager_global_ordinals"));
+                    iterator.remove();
+                }
+            }
+            return builder;
+        }
+    }
+
+    
+    /**
+     * Parses markdown-like syntax into plain text and AnnotationTokens with offsets for
+     * annotations found in texts
+     */
+    public static final class AnnotatedText {
+        public final String textPlusMarkup;
+        public final String textMinusMarkup;
+        List<AnnotationToken> annotations;
+        
+        // Format is markdown-like syntax for URLs eg:
+        //   "New mayor is [John Smith](type=person&value=John%20Smith) "
+        static Pattern markdownPattern = Pattern.compile("\\[([^\\]\\[]*)\\]\\(([^\\)\\(]*)\\)");  
+        
+        public static AnnotatedText parse (String textPlusMarkup) {
+            List<AnnotationToken> annotations =new ArrayList<>();
+            Matcher m = markdownPattern.matcher(textPlusMarkup);                
+            int lastPos = 0;
+            StringBuilder sb = new StringBuilder();
+            while(m.find()){
+                if(m.start() > lastPos){
+                    sb.append(textPlusMarkup.substring(lastPos, m.start()));
+                }
+                
+                int startOffset = sb.length();
+                int endOffset = sb.length() + m.group(1).length();
+                sb.append(m.group(1));
+                lastPos = m.end();
+                
+                String[] pairs = m.group(2).split("&");
+                String type = null;
+                String value = null;
+                for (String pair : pairs) {
+                    String[] kv = pair.split("=");
```

Comment:
Future stuff. Significant_text currently tends to offer muddled suggestions - the tokens `donald`, `trump` and `Donald Trump`.  They all generally relate to the same entity in the text of matching docs but it would be nice to be able to have an option in significant_text to take annotation tokens (e.g. type `Person`) in preference to any plain-text tokens it spans. It is, after all, a piece of the unstructured text where someone or something has understood exactly what the text is about and offered the canonical form of the entity.
This would be an optional choice of behaviour because when aiming for precision you'd only want the annotation Vs when aiming for recall you'd ask for both annotations and the underlying text.

To support this we'd need typed tokens.